### PR TITLE
Add simplified helm chart

### PR DESCRIPTION
**Pull request by S7evinK**
As discussed yesterday, a simplified version of [my helm](https://github.com/S7evinK/dendrite-helm) which deploys a monolith with internal NATS and an optionally enabled PostgreSQL server. If the PostgreSQL dependency is not enabled, a user specified connection string is constructed.

### CODE REVIEW

File: helm/dendrite/.helm-docs/database.gotmpl [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+You'll need to create the following databases before starting Dendrite (see [install.md](https://github.com/matrix-org/dendrite/blob/master/docs/INSTALL.md#configuration)):
+
+```postgres
+create database dendrite_federationapi;
+create database dendrite_mediaapi;
+create database dendrite_roomserver;
+create database dendrite_userapi_accounts;
+create database dendrite_keyserver;
+create database dendrite_syncapi;
```
Comment (kegsay) [inline helm/dendrite/.helm-docs/database.gotmpl:14]: We should not have this chart support polylith. One database only please.

File: helm/dendrite/templates/secrets.yaml [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+{{ end }}
+---
+apiVersion: v1
+kind: Secret
+type: Opaque
+metadata:
+  name: {{ .Release.Name }}-conf
+  namespace: {{ .Release.Namespace }}
+stringData:
+  dendrite.yaml: |
```
Comment (kegsay) [inline helm/dendrite/templates/secrets.yaml:36]: It's kind of annoying to put the entire config YAML as a secret, as opposed to the actually secret stuff. It makes casually browsing deployments much more difficult.

File: helm/dendrite/values.yaml [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+    # -- The TURN username
+    turn_username: ""
+    # -- The TURN password
+    turn_password: ""
+
+  rate_limiting:
+    # -- Enable rate limiting
+    enabled: true
+    # -- After how many requests a rate limit should be activated
+    threshold: 5
```
Comment (kegsay) [inline helm/dendrite/values.yaml:92]: Very low default.

File: helm/dendrite/values.yaml [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+# This can be set individually for each component, see mediaapi
+# @default -- sets some sane default values
+resources:
+  requests:
+    cpu: "80m"
+    memory: "64Mi"
+
+  limits:
+    cpu: "240m"
+    memory: "256Mi"
```
Comment (kegsay) [inline helm/dendrite/values.yaml:148]: Almost certainly too low for any sufficiently large deployment.

File: helm/dendrite/values.yaml [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+    method: crop
+  - width: 640
+    height: 480
+    method: scale
+
+syncapi:
+  # -- This option controls which HTTP header to inspect to find the real remote IP
+  # address of the client. This is likely required if Dendrite is running behind
+  # a reverse proxy server.
+  real_ip_header: X-Real-IP
```
Comment (kegsay) [inline helm/dendrite/values.yaml:212]: Does this work with the ingress reverse proxy?

**Review by kegsay — COMMENTED** [review]
Getting there, it mostly looks good.

How do we test the Helm chart currently?

File: helm/dendrite/templates/ingress.yaml [hunk 6]
```diff
@@ -1,28 +1,36 @@
...
-  name: {{ $component }}
+  name: {{ $fullName }}
   labels:
-    app: {{ $.Chart.Name }}
-    component: {{ $component }}
-  {{- with .Values.ingress.annotations }}
+    {{- include "dendrite.labels" . | nindent 4 }}
   annotations:
+    {{- if eq .Values.ingress.annotateFor "nginx" }}
+    nginx.ingress.kubernetes.io/upstream-hash-by: "$http_authorization"
```
Comment (kegsay) [inline helm/dendrite/templates/ingress.yaml:23]: Realistically we don't support this in Dendrite I think? I'd remove `annotateFor` entirely.

File: helm/dendrite/README.md [hunk 7]
```diff
@@ -72,12 +66,12 @@ Create a folder `appservices` and place your configurations in there.  The confi
 | configuration.metrics.basic_auth.password | string | `"metrics"` | HTTP basic authentication password |
 | configuration.metrics.basic_auth.user | string | `"metrics"` | HTTP basic authentication username |
 | configuration.metrics.enabled | bool | `false` | Whether or not Prometheus metrics are enabled. |
-| configuration.mscs | list | `[]` | Configuration for experimental MSC's. (Valid values are: msc2836 and msc2946) |
-| configuration.profiling.enabled | bool | `false` | Enable pprof |
+| configuration.mscs | list | `["msc2946"]` | Configuration for experimental MSC's. (Valid values are: msc2836 and msc2946) |
```
Comment (kegsay) [inline helm/dendrite/README.md:69]: I don't understand why we are inconsistent on key names. This is `configuration.mscs` as if `configuration` refers to the config YAML file, but then we have `clientapi.registration.registration_disabled` and others which lack the `configuration` prefix :S also `config` would be slightly shorter/nicer to type.

**Review by kegsay — CHANGES_REQUESTED** [review]
Before this can land, we need to:
 - lint the helm chart in CI.
 - test the helm chart works in minikube in CI.

File: helm/dendrite/README.md (hunk 7 above)
Comment (S7evinK) [reply helm/dendrite/README.md]: Updated, should be more consistent now.
The sections are based on the Dendrite config.

File: helm/dendrite/templates/secrets.yaml [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+    version: 2
+    global:
+      server_name: {{ .Values.global.server_name }}
+      private_key: /etc/dendrite/secrets/signing.key
+      key_id: ed25519:auto
+      key_validity_period: {{ .Values.global.key_validity_period | quote }}
+      database:
+        connection_string: {{ $connectionString }}?sslmode=disable
+        max_open_conns: {{ default 90 .Values.global.database.max_open_conns }}
```
Comment (genofire) [inline helm/dendrite/templates/secrets.yaml:45]: please put this defaults to `values.yaml` (so you do not need it here ...)

File: helm/dendrite/templates/deployment.yaml [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+    app: {{ $.Chart.Name }}
+spec:
+  selector:
+    matchLabels:
+      {{- include "dendrite.selectorLabels" . | nindent 6 }}
+  replicas: 1
+  template:
+    metadata:
+      labels:
+        {{- include "dendrite.selectorLabels" . | nindent 8 }}
```
Comment (genofire) [inline helm/dendrite/templates/deployment.yaml:18]:
```suggestion
        {{- include "dendrite.labels" . | nindent 8 }}
```

File: helm/dendrite/templates/deployment.yaml [hunk 10]
```diff
@@ -0,0 +1,97 @@
+{{ template "validate.config" . }}
+---
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  namespace: {{ $.Release.Namespace }}
+  name: {{ include "dendrite.fullname" . }}
+  labels:
+    app: {{ $.Chart.Name }}
```
Comment (genofire) [inline helm/dendrite/templates/deployment.yaml:9]: put on every resource the labels by your helper:
```suggestion
   {{- include "dendrite.labels" . | nindent 4 }}
```

File: helm/dendrite/templates/deployment.yaml [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+          secretName: {{ default (print .Release.Name "-signing-key") $.Values.signing_key.existingSecret | quote }}
+      {{- if (gt (len ($.Files.Glob "appservices/*")) 0) }}
+      - name:{{ .Release.Name }}-appservices
+        secret:
+          secretName: {{ .Release.Name }}-appservices-conf
+      {{- end }}
+      - name: {{ .Release.Name }}-jetstream
+        persistentVolumeClaim:
+          claimName: {{ default (print .Release.Name "-jetstream-pvc") $.Values.persistence.jetstream.existingClaim | quote }}
+      - name: {{ .Release.Name }}-media
```
Comment (genofire) [inline helm/dendrite/templates/deployment.yaml:35]: please replace everywhere `.Release.Name` with your helper
```suggestion
      - name: {{ include "dendrite.fullname" . }}-media
```

File: helm/dendrite/templates/deployment.yaml [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+  name: {{ include "dendrite.fullname" . }}
+  labels:
+    app: {{ $.Chart.Name }}
+spec:
+  selector:
+    matchLabels:
+      {{- include "dendrite.selectorLabels" . | nindent 6 }}
+  replicas: 1
+  template:
+    metadata:
```
Comment (genofire) [inline helm/dendrite/templates/deployment.yaml:16]: add additional label / annotations for restart pods without downtime if config file / secrets changes e.g.
```suggestion
    metadata:
      annotations:
        confighash-global: secret-{{ .Values.global | toYaml | sha256sum | trunc 32 }}
        confighash-clientapi: clientapi-{{ .Values.clientapi | toYaml | sha256sum | trunc 32 }}
        confighash-federationapi: federationapi-{{ .Values.federationapi | toYaml | sha256sum | trunc 32 }}
        confighash-mediaapi: mediaapi-{{ .Values.mediaapi | toYaml | sha256sum | trunc 32 }}
        confighash-syncapi: syncapi-{{ .Values.syncapi | toYaml | sha256sum | trunc 32 }}
```

File: helm/dendrite/templates/jobs.yaml [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+  name: {{ $name }}
+subjects:
+  - kind: ServiceAccount
+    name: {{ $name }}
+    namespace: {{ .Release.Namespace }}
+---
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: generate-signing-key
```
Comment (genofire) [inline helm/dendrite/templates/jobs.yaml:48]: nice <3

**Review by genofire — CHANGES_REQUESTED** [review]
Nice first version, here a few things i would change.

**Comment by S7evinK** [comment]
@genofire Thanks for your review, I've changed most of those, and ct still seems happy.

File: helm/dendrite/templates/secrets.yaml [hunk 14]
```diff
@@ -30,88 +30,63 @@ apiVersion: v1
...
       metrics:
-        {{- toYaml .Values.configuration.metrics | nindent 8 }}
+        {{- toYaml .Values.dendrite_config.global.metrics | nindent 8 }}
       dns_cache:
-        {{- toYaml .Values.configuration.dns_cache | nindent 8 }}
+        {{- toYaml .Values.dendrite_config.global.dns_cache | nindent 8 }}
     mscs:
       mscs:
-        {{- toYaml .Values.configuration.mscs | nindent 8 }}
+        {{- toYaml .Values.dendrite_config.mscs.mscs | nindent 8 }}
```
Comment (kegsay) [inline helm/dendrite/templates/secrets.yaml:71]: As discussed yesterday, please can we just drop-in `.Values.dendrite_config` directly into the value for this `stringData`? That way we don't need to update this file when config options change, and it forces us to put all defaullts in `Values.yml`.

**Review by kegsay — APPROVED** [review]
Other than comments, LGTM

**Pull request merged**